config/hostid/hostid.go

```diff
@@ -50,34 +67,151 @@ type Configuration struct {
-			return "", err
-		}
-		return *c.Value, nil
+		return &configResolver{value: c.Value}, nil
 	case EnvironmentResolver:
-		if c.EnvVarName == nil {
-			err := fmt.Errorf("missing host ID env var name using: resolver=%s",
-				string(c.Resolver))
-			return "", err
+		return &environmentResolver{envVarName: c.EnvVarName}, nil
+	case KeyValueFileResolver:
+		if c.KVFile == nil {
+			return nil, errors.New("KV file config cannot be nil")
 		}
-		v := os.Getenv(*c.EnvVarName)
-		if v == "" {
-			err := fmt.Errorf("missing host ID env var value using: resolver=%s, name=%s",
-				string(c.Resolver), *c.EnvVarName)
+		return &kvFile{
+			key:     c.KVFile.Key,
+			path:    c.KVFile.Path,
+			timeout: c.KVFile.Timeout,
+		}, nil
+	}
+	return nil, fmt.Errorf("unknown host ID resolver: resolver=%s",
+		string(c.Resolver))
+}
+
+// Resolve returns the resolved host ID given the configuration.
+func (c Configuration) Resolve() (string, error) {
+	r, err := c.resolver()
+	if err != nil {
+		return "", err
+	}
+	return r.ID()
+}
+
+type hostnameResolver struct{}
+
+func (hostnameResolver) ID() (string, error) {
+	return os.Hostname()
+}
+
+type configResolver struct {
+	value *string
+}
+
+func (c *configResolver) ID() (string, error) {
+	if c.value == nil {
+		err := fmt.Errorf("missing host ID using: resolver=%s", string(ConfigResolver))
+		return "", err
+	}
+	return *c.value, nil
+}
+
+type environmentResolver struct {
+	envVarName *string
+}
+
+func (c *environmentResolver) ID() (string, error) {
+	if c.envVarName == nil {
+		err := fmt.Errorf("missing host ID env var name using: resolver=%s",
+			string(EnvironmentResolver))
+		return "", err
+	}
+	v := os.Getenv(*c.envVarName)
+	if v == "" {
+		err := fmt.Errorf("missing host ID env var value using: resolver=%s, name=%s",
+			string(EnvironmentResolver), *c.envVarName)
+		return "", err
+	}
+	return v, nil
+}
+
+type kvFile struct {
+	path     string
+	key      string
+	interval time.Duration
+	timeout  time.Duration
+}
+
+// ID attempts to parse an ID from a KV config file. It will optionally wait a
+// timeout to find the value, as in some environments the file may be
+// dynamically generated from external metadata and not immediately available
+// when the instance starts up.
+func (c *kvFile) ID() (string, error) {
+	checkF := func() (string, error) {
+		f, err := os.Open(c.path)
+		if err != nil {
 			return "", err
 		}
-		return v, nil
+
+		return c.parseKV(f)
 	}
-	return "", fmt.Errorf("unknown host ID resolver: resolver=%s",
-		string(c.Resolver))
+
+	if c.timeout == 0 {
+		return checkF()
+	}
+
+	interval := c.interval
+	if interval == 0 {
+		interval = defaultFileCheckInterval
+	}
+
+	startT := time.Now()
+	for time.Since(startT) < c.timeout {
+		v, err := checkF()
+		if err == nil {
+			return v, nil
+		}
+
+		time.Sleep(interval)
+	}
+
+	return "", fmt.Errorf("did not find value in %s within %s", c.path, c.timeout)
+}
+
+func (c *kvFile) parseKV(r io.Reader) (string, error) {
```

Comment:
Any reason not to use JSON or YAML for this?  Seems a little strange to implement parser yourself.  

You could also just make it a plain text file, with nothing else in it. (so `ioutil.ReadAll(...)`)

Reply:
Unfortunately we're limited to the format that Kubernetes spits out the annotations in, but I just realized that we can configure just 1 annotation field to be dumped and then could read the file's content as you mentioned. Will update.